futures-core/src/future/mod.rs

```diff
@@ -100,75 +100,83 @@ pub trait Future {
+        (**self).poll(cx)
     }
 }
 
 if_std! {
-    use std::boxed::{Box, PinBox};
-
-    impl<'a, F: ?Sized + Future> Future for Box<F> {
-        type Output = F::Output;
+    impl<F: ?Sized + Future> Future for ::std::boxed::Box<F> {
+        type Item = F::Item;
+        type Error = F::Error;
 
-        fn poll(mut self: Pin<Self>, cx: &mut task::Context) -> Poll<Self::Output> {
-            unsafe { pinned_deref!(self).poll(cx) }
+        fn poll(&mut self, cx: &mut task::Context) -> PollResult<Self::Item, Self::Error> {
+            (**self).poll(cx)
         }
     }
 
-    impl<'a, F: ?Sized + Future> Future for PinBox<F> {
-        type Output = F::Output;
+    #[cfg(feature = "nightly")]
+    impl<F: ?Sized + Future> Future for ::std::boxed::PinBox<F> {
+        type Item = F::Item;
+        type Error = F::Error;
 
-        fn poll(mut self: Pin<Self>, cx: &mut task::Context) -> Poll<Self::Output> {
-            unsafe {
-                let this = PinBox::get_mut(Pin::get_mut(&mut self));
-                let re_pinned = Pin::new_unchecked(this);
-                re_pinned.poll(cx)
-            }
+        fn poll(&mut self, cx: &mut task::Context) -> PollResult<Self::Item, Self::Error> {
+            unsafe { ::core::mem::Pin::get_mut(&mut self.as_pin()).poll(cx) }
         }
     }
 
-    impl<'a, F: Future> Future for ::std::panic::AssertUnwindSafe<F> {
-        type Output = F::Output;
+    impl<F: Future> Future for ::std::panic::AssertUnwindSafe<F> {
+        type Item = F::Item;
+        type Error = F::Error;
 
-        fn poll(mut self: Pin<Self>, cx: &mut task::Context) -> Poll<Self::Output> {
-            unsafe { pinned_field!(self, 0).poll(cx) }
+        fn poll(&mut self, cx: &mut task::Context) -> PollResult<F::Item, F::Error> {
+            self.0.poll(cx)
         }
     }
 }
 
-/// A future that is immediately ready with a value
-#[derive(Debug, Clone)]
-#[must_use = "futures do nothing unless polled"]
-pub struct ReadyFuture<T>(Option<T>);
-
-unsafe impl<T> Unpin for ReadyFuture<T> {}
-
-impl<T> Future for ReadyFuture<T> {
-    type Output = T;
-
-    fn poll(mut self: Pin<Self>, _cx: &mut task::Context) -> Poll<T> {
-        Poll::Ready(self.0.take().unwrap())
-    }
+/// Types that can be converted into a future.
+///
+/// This trait is very similar to the `IntoIterator` trait.
+pub trait IntoFuture {
+    /// The future that this type can be converted into.
+    type Future: Future<Item=Self::Item, Error=Self::Error>;
+
+    /// The item that the future may resolve with.
+    type Item;
+    /// The error that the future may resolve with.
+    type Error;
+
+    /// Consumes this object and produces a future.
+    fn into_future(self) -> Self::Future;
 }
 
-/// Create a future that is immediately ready with a value.
-pub fn ready<T>(t: T) -> ReadyFuture<T> {
-    ReadyFuture(Some(t))
+impl<F> IntoFuture for F where F: Future {
```

Comment:
I think you also want an `impl<A> IntoFuture for A where A: AsyncResult + Unpin { ... }`